### ISSUE TITLE
Domains: Able to reach checkout with Punycode version of IDN

### ISSUE THREAD

**Issue by SiobhyB** [Automattic/wp-calypso#380]
When attempting to register a domain, siobhánbamber.com, via https://wordpress.com/domains/add, I'm blocked by an error notice informing me that the domain isn't valid. This is expected as WordPress.com [doesn't support the registration or mapping of IDNs (Internationalized Domain Names)](https://en.support.wordpress.com/map-existing-domain/#internationalized-domain-names-idns).

<img width="784" alt="screen shot 2015-11-21 at 12 55 17" src="https://cloud.githubusercontent.com/assets/2998162/11318546/2aa323f2-904f-11e5-850a-da0990f433e4.png">

However, when attempting to register xn--siobhn-ttabamber.com (which is the [IDNA encoded](https://en.wikipedia.org/wiki/Internationalized_domain_name) or Punycode version) of siobhánbamber.com then I was able to reach checkout.

<img width="753" alt="screen shot 2015-11-21 at 13 01 05" src="https://cloud.githubusercontent.com/assets/2998162/11318573/f64b7e96-904f-11e5-9912-fbeae6c5834e.png">

The registration technically failed as xn--siobhn-ttabamber.com isn't supported at WordPress.com, however, no notice of the failure was given throughout the registration process.

**Comment by umurkontaci** [comment]
/cc: @aidvu since this better be done in the backend

**Comment by umurkontaci** [comment]
@SiobhyB Turns out there's already a check for this type for user accounts.

**Comment by mahangu** [comment]
Just reopening this ticket because I am able to reproduce this error now from Tortuga:

1) Go to https://wordpress.com/domains/

2) Search for any punycode domain like `xn--8zc3ctb0d.com`

![screen shot 2017-05-13 at 6 31 06 pm](https://cloud.githubusercontent.com/assets/10125810/26025677/8e7bd4e6-380a-11e7-9a01-34418dd47814.png)

3) Select it and proceed to checkout:

![screen shot 2017-05-13 at 6 31 22 pm](https://cloud.githubusercontent.com/assets/10125810/26025678/95f6bef2-380a-11e7-9170-3311e6307dd5.png)

![screen shot 2017-05-13 at 6 31 13 pm](https://cloud.githubusercontent.com/assets/10125810/26025684/a78d7bc4-380a-11e7-8b77-ab46fff6124c.png)

4) Checkout, and observe how it throws this error:

![screen shot 2017-05-13 at 6 34 19 pm](https://cloud.githubusercontent.com/assets/10125810/26025709/0d70d9ae-380b-11e7-8120-35a60c5d85d0.png)

Ideally, we should prevent the user from ever being able to add the punycode domain to their cart, like we do from `/domains/add/`.

Please let me know if I should file this on the Tortuga GH repo instead. Thank you!

**Comment by umurkontaci** [comment]
@mahangu can you confirm with a regular user account?

**Comment by mahangu** [comment]
@umurkontaci yes, I am able to reproduce it on a regular account, unproxied.

![screen shot 2017-05-16 at 11 24 10 am](https://cloud.githubusercontent.com/assets/10125810/26091648/3607c47e-3a2a-11e7-85e7-05a122a9655b.png)

**Comment by klimeryk** [comment]
Seems like there are two aspects of this - if you search for a direct match, like `xn--8zc3ctb.com`, you'll be able to add it to cart. But the actual domain should be `මහඟ.com`, which you can't add to cart, because of the availability check that runs after you try to add it to cart.
We should also prevent Domainsbot from suggesting IDN domains, because right now it does :/

**Comment by klimeryk** [comment]
This is not yet fixed - the patch helped with hiding this issue for Tortuga users, but if someone searches for a domain in punycode format it will still allow it to register it.

**Comment by klimeryk** [comment]
cc @Automattic/i18n - I was wondering what's your take on internationalized domain names (aka IDN)? My guess would be that they are popular in some countries (with a non-latin alphabet perhaps), but would love to hear your expertise, so that we can better prioritize this issue.

**Comment by naokomc** [comment]
Japanese domains are not super common but not too rare either.

A Japanese page on [Wikipedia says](https://ja.wikipedia.org/wiki/日本語ドメイン名) 120K out of 900K .JP domains are Japanese domains, but I don't feel I see them very often. Maybe people register them but use alphabet version as the primary one.

FYI - according to https://jprs.jp/faq/use/ , ドメイン名例.JP and XN--ECKWD4C7CU47R2WF.JP (Punycode version) can be used as test domains, similar to example.com.

**Comment by deBhal** [comment]
I don't think we've done any real investigation.

Tucows says they support IDNs in some languages for some tlds, which sounds "pretty fun", but we're in a good place to handle that sort of thing now (I don't know how hard that "some languages" would be).

The big question mark is the places where we've made broken assumptions. For example, I saw a thread where a publicize link was broken in facebook because of the encoding, and videopress has it's own `is_valid_domain()` that would reject IDNs whether or not they're encoded.

There are also some security questions around using visually similar characters to nefarious ends.

We don't even support mapping IDNs right now, but it looks like there are some tricks kicking around to make it work sometimes, so it's not all bad :)

We need some idea of the value here to compare the effort to. How valuable are IDNs to users?

**Comment by klimeryk** [comment]
Thanks for chiming in, @naokomc and @deBhal 🙇 Maybe let's start then with just supporting mapping such domains (it _is_ the prerequisite for support registering such domains anyway) and see how popular those are. Meaning, if someone searches for an IDN, let's direct them to map the domain (if it's taken) or show them the usual green nudge to buy it if a non-IDN version is available (if it's possible to easily convert to such - easy for stuff like żłóć, but Japanese alphabet not so much ;)). Sounds good? Any volunteers? 😁

**Comment by aidvu** [comment]
We would have to store them in the non-IDN representation in the DB, because collation/charset, so some mumbo-jumbo would be required encoding/decoding them every time. 

Would be nice to have it, yeah, but needs a bit more research on P2. :)

**Comment by klimeryk** [comment]
Right - probably in the Punycode format. And that seems the biggest technical hurdle here - it will introduce a different "raw" domain and a different one for display purposes.
